services/src/datasets/storage.rs

```diff
@@ -0,0 +1,188 @@
+        }
+    }
+}
+
+impl UserInput for ImportDataSet {
+    fn validate(&self) -> Result<()> {
+        // TODO
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSetProviderListing {
+    pub id: DataSetProviderId,
+    pub name: String,
+    pub description: String,
+    // more meta data (number of data sets, ...)
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum AddDataSetProvider {
+    AddMockDataSetProvider(AddMockDataSetProvider),
+    // TODO: geo catalog, wcs, ...
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct AddMockDataSetProvider {
+    pub data_sets: Vec<DataSet>,
+}
+
+impl UserInput for AddDataSetProvider {
+    fn validate(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSetProviderListOptions {
+    // TODO: filter
+    pub offset: u32,
+    pub limit: u32,
+}
+
+impl UserInput for DataSetProviderListOptions {
+    fn validate(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+#[allow(clippy::large_enum_variant)] // TODO: box?
+pub enum DataSetLoadingInfo {
+    Raster(RasterLoadingInfo),
+    Vector(VectorLoadingInfo),
+}
+
+pub struct GdalLoadingInfo {
+    pub file: PathBuf,
+}
+
+pub type RasterLoadingInfo = GdalLoadingInfo;
+
+#[allow(clippy::large_enum_variant)] // TODO: box?
+pub enum VectorLoadingInfo {
+    Mock(MockDataSetDataSourceLoadingInfo),
+    Ogr(OgrSourceDataset),
+}
+
+/// Handling of data sets provided by geo engine internally, staged and by external providers
+#[async_trait]
+pub trait DataSetDB: DataSetProvider + Send + Sync {
+    /// Add an external data set `provider` by `user`
+    // TODO: require special privilege to be able to add external data set provider and to access external data in general
+    async fn add_data_set_provider(
+        &mut self,
+        user: UserId,
+        provider: Validated<AddDataSetProvider>,
+    ) -> Result<DataSetProviderId>;
+
+    // TODO: share data set provider/manage permissions
+
+    /// List available providers for `user` filtered by `options`
+    async fn list_data_set_providers(
+        &self,
+        user: UserId,
+        options: Validated<DataSetProviderListOptions>,
+    ) -> Result<Vec<DataSetProviderListing>>;
+
+    /// Get data set `provider` for `user`
+    async fn data_set_provider(
+        &self,
+        user: UserId,
+        provider: DataSetProviderId,
+    ) -> Result<&dyn DataSetProvider>;
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub enum DataSetPermission {
+    Read,
+    Write,
+    Owner,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub struct UserDataSetPermission {
+    pub user: UserId,
+    pub data_set: InternalDataSetId,
+    pub permission: DataSetPermission,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub enum DataSetProviderPermission {
+    Read,
+    Write,
+    Owner,
+}
```

Comment:
This also looks redundant, I guess we should think about creating a util for that.

Reply:
we could think about a macro once it's all working